{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "800ab3a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import io\n",
    "from typing import Tuple, Union, Iterator, Generator\n",
    "\n",
    "\n",
    "class Parser:\n",
    "    \"\"\"\n",
    "    Base Class for Parsing Algorithms\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, filename: str):\n",
    "        \"\"\"\n",
    "        Initialization to be shared by all inherited classes\n",
    "\n",
    "        # What does the `__init__` method do?\n",
    "            This method will be called immediately upon creating an object. It's a useful\n",
    "            way to assign baseline attributes of the class (in this case making the filename\n",
    "            accessible by all methods) but also to run preliminary code or assertions (like\n",
    "            checking to see if the file exists at all!)\n",
    "\n",
    "        # Should I ever call the `__init__` method?\n",
    "            Like most hidden methods (the double underscored names) this is not generally something\n",
    "            you call from the outside of a class. However, if you need to specify a different\n",
    "            `__init__` method for a subclass you will need to call this with the `Super` keyword. We\n",
    "            won't get into this now, but if you are interested feel free to reach out to the TAs or\n",
    "            check out the documentation on the `Super` keyword.\n",
    "        \"\"\"\n",
    "        self.filename = filename\n",
    "        self.store = True\n",
    "        self._sequences = None\n",
    "\n",
    "    def get_record(\n",
    "        self, f_obj: io.TextIOWrapper\n",
    "    ) -> Union[Tuple[str, str], Tuple[str, str, str]]:\n",
    "        \"\"\"\n",
    "        Returns a sequencing record that will either be a tuple of two strings (header, sequence)\n",
    "        or a tuple of three strings (header, sequence, quality).\n",
    "\n",
    "        # What's the deal with calling a method by almost the same name?\n",
    "            it is common in python to see a public method calling a hidden method\n",
    "            with a similar name. Both of these are accessible to a user (nothing is truly hidden in python)\n",
    "            but it is a useful way to separate out Class and SubClass specific behavior.\n",
    "\n",
    "            In this case, we know that the function will return either a tuple of 2 or a tuple of 3.\n",
    "            But it is up to the subclass method to define what tuple it will return.\n",
    "\n",
    "        # Do I need to do this with all my classes?\n",
    "            Absolutely not. But we want to show you some things you will see often when reading python code\n",
    "            and give an explanation for why certain practices exist in the language.\n",
    "\n",
    "        \"\"\"\n",
    "        return self._get_record(f_obj)\n",
    "\n",
    "    def __iter__(self):\n",
    "        \"\"\"\n",
    "        This is an overriding of the Base Class Iterable function. All classes in python\n",
    "        have this function, but it is not implemented for all classes in python.\n",
    "\n",
    "        # Note on the `__iter__` method\n",
    "            Generally one doesn't call this method directly as `obj.__iter__()`. Instead it\n",
    "            lets you use the object itself as an iterable. This is really useful in OOP because it\n",
    "            allows you to represent and use iterable objects very cleanly. You still can call this\n",
    "            method directly, but it really takes the fun out of python...\n",
    "\n",
    "            ## How to use the `__iter__` method\n",
    "            ```\n",
    "            parser_obj = Parser(filename)\n",
    "            for record in parser_obj:\n",
    "              # do something\n",
    "            ```\n",
    "\n",
    "        # Why you should care about generators\n",
    "\n",
    "            The expected behavior of this function is to create a generator which will lazily load\n",
    "            the next item in its queue. These are very useful for many bioinformatic tools where you\n",
    "            don't need everything loaded at once and instead are interested in interacting with the\n",
    "            stream (i.e. you need every value once and won't need it again after you use it). This saves\n",
    "            quite a bit of memory, especially when you are working with billions of sequences and don't\n",
    "            need to keep all of them in memory.\n",
    "\n",
    "        # Distinction between generator functions and other functions\n",
    "\n",
    "            instead of returning a value with the keyword `return`\n",
    "            a generator must return a value with the keyword `yield`.\n",
    "\n",
    "            This `yield` keyword will not shortcut the loop it is nested in like a return will\n",
    "            and instead will pause the loop until the object is taken from it.\n",
    "        \"\"\"\n",
    "\n",
    "        # The proper way to open a file for reading and writing in python3 is to use the `with` / `as` keywords.\n",
    "        # and keep the I/O within the nested code block. This will save you from some really nasty bugs that\n",
    "        # sometimes close a file before everything you expect to be written/read is written/read.\n",
    "        #\n",
    "        # the interpretation of the following code is that for the lifetime of the filebuffer\n",
    "        # returned by the `open` function it will be accessible as the variable `f_obj`\n",
    "\n",
    "        nseq = 0\n",
    "        with open(self.filename, \"r\") as f_obj:\n",
    "            rec = self.get_record(\n",
    "                f_obj\n",
    "            )  # will be a generator that yields tuples of strings\n",
    "\n",
    "            for seq in rec:\n",
    "                yield seq\n",
    "                nseq += 1\n",
    "            self.store = False\n",
    "\n",
    "            if nseq == 0:\n",
    "                raise ValueError(f\"File ({self.filename}) had 0 lines.\")\n",
    "\n",
    "        # another way to do this with the original construction: \n",
    "        #    while True:\n",
    "        #        rec = self.get_record(f_obj)\n",
    "        #        yield rec \n",
    "        # is to implement for the FastaParser/FastqParser subclasses' _get_record\n",
    "        # functionality where you get (with `next(f_obj)`, for example) two lines\n",
    "        # at a time for the FastaParser and simply assume the first is the \n",
    "        # header and the second is the sequence, or something similar.\n",
    "\n",
    "    def _get_record(\n",
    "        self, f_obj: io.TextIOWrapper\n",
    "    ) -> Iterator[Union[Tuple[str, str], Tuple[str, str, str]]]:\n",
    "        \"\"\"\n",
    "        a method to be overridden by inherited classes.\n",
    "        \"\"\"\n",
    "        raise NotImplementedError(\n",
    "            \"\"\"\n",
    "                This function is not meant to be called by the Parser Class.\n",
    "                It is expected to be overridden by `FastaParser` and `FastqParser`\"\"\"\n",
    "        )\n",
    "\n",
    "\n",
    "class FastaParser(Parser):\n",
    "    \"\"\"\n",
    "    Fasta Specific Parsing\n",
    "    \"\"\"\n",
    "\n",
    "    def _get_record(self, f_obj: io.TextIOWrapper) -> Iterator[Tuple[str, str]]:\n",
    "        \"\"\"\n",
    "        returns the next fasta record\n",
    "        \"\"\"\n",
    "\n",
    "        seq_name = None\n",
    "\n",
    "        for idx, line in enumerate(f_obj):\n",
    "            line = line.strip()\n",
    "            if line == \"\":\n",
    "                raise ValueError(f\"Got an empty line for {f_obj.name} @ line {idx+1}\")\n",
    "            if line.startswith(\">\"):\n",
    "                seq_name = line[1:]\n",
    "                continue\n",
    "\n",
    "            yield (seq_name, line)\n",
    "\n",
    "\n",
    "class FastqParser(Parser):\n",
    "    \"\"\"\n",
    "    Fastq Specific Parsing\n",
    "    \"\"\"\n",
    "\n",
    "    def _get_record(\n",
    "        self, f_obj: io.TextIOWrapper\n",
    "    ) -> Generator[Tuple[str, str, str], None, None]:\n",
    "        \"\"\"\n",
    "        returns the next fastq record\n",
    "        \"\"\"\n",
    "        read_qual = True\n",
    "        seq_name = None\n",
    "        seq = None\n",
    "\n",
    "        for idx, line in enumerate(f_obj):\n",
    "            line = line.strip()\n",
    "            if line == \"\":\n",
    "                raise ValueError(f\"Got an empty line for {f_obj.name} @ line {idx+1}\")\n",
    "            if line == \"+\":\n",
    "                continue  # skip this line\n",
    "\n",
    "            if line.startswith(\"@\"):  # if its a header line, we'll store it\n",
    "                seq_name = line[1:]\n",
    "                continue\n",
    "\n",
    "            if (\n",
    "                read_qual is True\n",
    "            ):  # if read_qual is True, then we'll assume the line is a sequence\n",
    "                seq = line\n",
    "                read_qual = False\n",
    "            else:\n",
    "                # we assume that quality will always be after the seq, so if we get here and read_qual is False then we can just return the tuple\n",
    "                yield (seq_name, seq, line)  # line here is the quality string\n",
    "                read_qual = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "064532b8",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "File (/Users/brysonchoy/Documents/bmi203/HW1-FAST-AQ-Parser-BrysonChoy/tests/empty.fa) had 0 lines.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Input \u001b[0;32mIn [10]\u001b[0m, in \u001b[0;36m<cell line: 69>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     62\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m     63\u001b[0m \u001b[38;5;124;03mWhen executing a python script from the command line there will\u001b[39;00m\n\u001b[1;32m     64\u001b[0m \u001b[38;5;124;03malways be a hidden variable `__name__` set to the value `__main__`.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     67\u001b[0m \u001b[38;5;124;03mthe following if statement\u001b[39;00m\n\u001b[1;32m     68\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m     69\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;18m__name__\u001b[39m \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m__main__\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m---> 70\u001b[0m     \u001b[43mmain\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "Input \u001b[0;32mIn [10]\u001b[0m, in \u001b[0;36mmain\u001b[0;34m()\u001b[0m\n\u001b[1;32m     24\u001b[0m delim \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m delim \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     26\u001b[0m \u001b[38;5;66;03m#fasta_file = fasta_file.name\u001b[39;00m\n\u001b[1;32m     27\u001b[0m \u001b[38;5;66;03m#fastq_file = fastq_file.name\u001b[39;00m\n\u001b[1;32m     28\u001b[0m \n\u001b[1;32m     29\u001b[0m \u001b[38;5;66;03m#if print_delim:\u001b[39;00m\n\u001b[1;32m     30\u001b[0m \u001b[38;5;66;03m#    print(delim + f\"Transcribing FASTA: {fasta_file}\\n\")\u001b[39;00m\n\u001b[0;32m---> 32\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m seq_name, seq \u001b[38;5;129;01min\u001b[39;00m fasta_parser:\n\u001b[1;32m     33\u001b[0m     \u001b[38;5;28mprint\u001b[39m(seq_name, transcribe(seq))\n\u001b[1;32m     35\u001b[0m \u001b[38;5;66;03m# For each record of FastqParser, Transcribe the sequence\u001b[39;00m\n\u001b[1;32m     36\u001b[0m \u001b[38;5;66;03m# and print it to console\u001b[39;00m\n",
      "File \u001b[0;32m~/Documents/bmi203/HW1-FAST-AQ-Parser-BrysonChoy/seqparser/parse.py:108\u001b[0m, in \u001b[0;36mParser.__iter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    105\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstore \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m    107\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m nseq \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m--> 108\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFile (\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfilename\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m) had 0 lines.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mValueError\u001b[0m: File (/Users/brysonchoy/Documents/bmi203/HW1-FAST-AQ-Parser-BrysonChoy/tests/empty.fa) had 0 lines."
     ]
    }
   ],
   "source": [
    "from seqparser import FastaParser, FastqParser, transcribe, reverse_transcribe\n",
    "import pathlib\n",
    "\n",
    "\n",
    "def main():\n",
    "    \"\"\"\n",
    "    The main function\n",
    "    \"\"\"\n",
    "    #data_dir = pathlib.Path(__file__).resolve().parent / \"data\"\n",
    "    #fasta_file = data_dir / \"test.fa\"\n",
    "    #fastq_file = data_dir / \"test.fq\"\n",
    "    fasta_file = \"/Users/brysonchoy/Documents/bmi203/HW1-FAST-AQ-Parser-BrysonChoy/tests/empty.fa\"\n",
    "\n",
    "    # Create instance of FastaParser\n",
    "    # Create instance of FastqParser\n",
    "\n",
    "    fasta_parser = FastaParser(fasta_file)\n",
    "    #fastq_parser = FastqParser(fastq_file)\n",
    "\n",
    "    # For each record of FastaParser, Transcribe the sequence\n",
    "    # and print it to console\n",
    "\n",
    "    delim = \"-\" * 50\n",
    "    delim = \"\\n\" + delim + \"\\n\"\n",
    "\n",
    "    #fasta_file = fasta_file.name\n",
    "    #fastq_file = fastq_file.name\n",
    "    \n",
    "    #if print_delim:\n",
    "    #    print(delim + f\"Transcribing FASTA: {fasta_file}\\n\")\n",
    "\n",
    "    for seq_name, seq in fasta_parser:\n",
    "        print(seq_name, transcribe(seq))\n",
    "\n",
    "    # For each record of FastqParser, Transcribe the sequence\n",
    "    # and print it to console\n",
    "\n",
    "    if print_delim:\n",
    "        print(delim + f\"Transcribing FASTQ: {fastq_file}\\n\")\n",
    "\n",
    "    for seq_name, seq, quality in fastq_parser:\n",
    "        print(seq_name, transcribe(seq))\n",
    "\n",
    "    # For each record of FastaParser, Reverse Transcribe the sequence\n",
    "    # and print it to console\n",
    "\n",
    "    if print_delim:\n",
    "        print(delim + f\"Reverse transcribing FASTA: {fasta_file}\\n\")\n",
    "\n",
    "    for seq_name, seq in fasta_parser:\n",
    "        print(seq_name, reverse_transcribe(seq))\n",
    "\n",
    "    # For each record of FastqParser, Reverse Transcribe the sequence\n",
    "    # and print it to console\n",
    "    if print_delim:\n",
    "        print(delim + f\"Reverse transcribing FASTQ: {fastq_file}\\n\")\n",
    "\n",
    "    for seq_name, seq, quality in fastq_parser:\n",
    "        print(seq_name, reverse_transcribe(seq))\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "When executing a python script from the command line there will\n",
    "always be a hidden variable `__name__` set to the value `__main__`.\n",
    "\n",
    "Since this is guaranteed you can execute your `main` function with\n",
    "the following if statement\n",
    "\"\"\"\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05243397",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1560f6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
